### PR TITLE
Allow TLS_RSA* ciphers in JDisc with Jetty 9.4.12+

### PR DESCRIPTION
**Pull request by bjorncs**
Jetty 9.4.12+ disables all TLS_RSA ciphers by default
(https://github.com/eclipse/jetty.project/issues/2807).

FYI @hakonhall @baldersheim

### CODE REVIEW

**Comment by bjorncs** [comment]
Note: Older TLS_RSA* ciphers (e.g using SHA/SHA1) will still be disabled. This change will only enable those ciphers that were enabled in Jetty 9.4.11.

**Review by freva — APPROVED** [review]

File: jdisc_http_service/src/main/java/com/yahoo/jdisc/http/server/jetty/ConnectorFactory.java [hunk 1]
```diff
@@ -106,6 +108,13 @@ private SslConnectionFactory newSslConnectionFactory() {
             factory.setSecureRandomAlgorithm(sslConfig.prng());
         }
 
+        // NOTE: ^TLS_RSA_.*$ ciphers are disabled by default in Jetty 9.4.12+ (https://github.com/eclipse/jetty.project/issues/2807)
+        // JDisc will allow these ciphers by default to support older clients (e.g. Java 8u60 and curl 7.29.0)
+        String[] excludedCiphersWithoutTlsRsaExclusion = Arrays.stream(factory.getExcludeCipherSuites())
+                .filter(cipher -> !cipher.equals("^TLS_RSA_.*$"))
+                .toArray(String[]::new);
+        factory.setExcludeProtocols(excludedCiphersWithoutTlsRsaExclusion);
```
Comment (hmusum) [inline jdisc_http_service/src/main/java/com/yahoo/jdisc/http/server/jetty/ConnectorFactory.java]: Why is setExcludeProtocols() used and not setExcludeCipherSuites()? Am I missing something?

Comment (freva) [reply jdisc_http_service/src/main/java/com/yahoo/jdisc/http/server/jetty/ConnectorFactory.java]: Hmm, good point. Looks like it should be cipher suites: https://github.com/eclipse/jetty.project/pull/2855/files

Comment (bjorncs) [reply jdisc_http_service/src/main/java/com/yahoo/jdisc/http/server/jetty/ConnectorFactory.java]: It's a typo, it should have been `setExcludeCipherSuites()`

**Comment by joakime** [comment]
Note: this exclusion is already present in the JVM on Java 11+ (your ConnectorFactory hack is unlikely to work on Java 11)

Might want to keep an eye on eclipse/jetty.project#2889 as well.

**Comment by bjorncs** [comment]
@joakime Thanks for letting us know. We discovered that some older versions of the JDK (e.g 1.8.0_60) failed to talk TLS with Jetty 9.4.12 without re-enabling the TLS_RSA ciphers.

**Comment by joakime** [comment]
The most common reasons the older clients cannot connect is the lack of Cipher Suites in common, or that TLS/1.2 isn't enabled.

For the lack of Cipher Suites in common, often just installing the JCE Unlimited Strength extensions on those JDKs is sufficient.
See https://www.oracle.com/technetwork/java/javase/downloads/jce8-download-2133166.html

For TLS/1.2 not being enabled, often that just requires a JVM command line option of `-Dhttps.protocols=SSLv3,TLSv1,TLSv1.1,TLSv1.2`

This is just general advice, I really don't know the nuances of the client implementations that you are referring to.

But I do want to leave you with a question for your project to think about (no need to answer here)
Are you sure you want to downgrade _all_ of the security on your server just to support old clients?

If the answer is yes, do your users a favor and let them know that your secure connection configuration is favoring support of older clients over modern security recommendations.

Know that other projects are also making the same decisions as yours is, so this isn't unique problem.

However, many of the other projects also put this kind of decision behind a configuration option that deployments of their servers can choose to use.  
An option that basically flips the decision back to use modern security recommendations intentionally breaking old client compatibility.

**Comment by bratseth** [comment]
@joakime excellent point, thanks for bringing it up.

**Pull request merged**